Add support for interface implementing interface

filosganga opened the pull request:
This is an initial support for #567 and supersede #597 that is based on very old base

The rendering has been a bit challenging and I took a non-standard path there

yanns commented on modules/core/src/main/scala/sangria/introspection/package.scala:
```diff
@@ -274,8 +274,8 @@ package object introspection {
           resolve = _.value._2 match {
             case t: ObjectType[_, _] =>
               Some(t.allInterfaces.asInstanceOf[Vector[Type]].map(true -> _))
-            case _: InterfaceType[_, _] =>
-              Some(Vector.empty)
+            case t: InterfaceType[_, _] =>
+              Some(t.interfaces.asInstanceOf[List[Type]].map(true -> _))
```
Both cases can be covered by `ObjectLikeType`.

filosganga replied:
Yep I have just re-read the spec, this is wrong, the transitively implemented interfaces should be ALL added

filosganga replied:
Fixed

yanns replied:
💯

yanns reviewed (commented):
Very cool to see progress here!

filosganga commented:
@yanns This is pretty much done but there are a couple of missing validations that I don't know how to handle:

**All the implementing fields must be declare**
When a type or an interface implement another interface, all the implementing fields must be declared. This is not currently validated even without interfaces implementing interfaces, so can it be addressed in a follow-up PR?

**The implementation hierarchy MUST not include loops**
I have managed to check for interfaces implementing themselves, however is very complex to spot issues like this:

````graphql
interface Foo implement Baz {
  foo: String
}

interface Bar implement Foo{
  bar: String
  foo: String
}

interface Baz implement Bar & Foo {
  baz: String
  bar: String
  foo: String
}
````

Because the materializer/builder goes in SO when trying to resolve the types. Is there a way to validate the ast before running the materializer/builder?

Cheers

yanns commented:
> **All the implementing fields must be declare** When a type or an interface implement another interface, all the implementing fields must be declared. This is not currently validated even without interfaces implementing interfaces, so can it be addressed in a follow-up PR?

Yes

> **The implementation hierarchy MUST not include loops** I have managed to check for interfaces implementing themselves, however is very complex to spot issues like this:
> 
> ```graphql
> interface Foo implement Baz {
>   foo: String
> }
> 
> interface Bar implement Foo{
>   bar: String
>   foo: String
> }
> 
> interface Baz implement Bar & Foo {
>   baz: String
>   bar: String
>   foo: String
> }
> ```
> 
> Because the materializer/builder goes in SO when trying to resolve the types. Is there a way to validate the ast before running the materializer/builder?

I'm not sure. I'm thinking of https://github.com/sangria-graphql/sangria/pull/966/files were some validation is done on the AST directly. Does this help?

filosganga commented:
> > **All the implementing fields must be declare** When a type or an interface implement another interface, all the implementing fields must be declared. This is not currently validated even without interfaces implementing interfaces, so can it be addressed in a follow-up PR?
> 
> Yes
> 
> > **The implementation hierarchy MUST not include loops** I have managed to check for interfaces implementing themselves, however is very complex to spot issues like this:
> > ```graphql
> > interface Foo implement Baz {
> >   foo: String
> > }
> > 
> > interface Bar implement Foo{
> >   bar: String
> >   foo: String
> > }
> > 
> > interface Baz implement Bar & Foo {
> >   baz: String
> >   bar: String
> >   foo: String
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Because the materializer/builder goes in SO when trying to resolve the types. Is there a way to validate the ast before running the materializer/builder?
> 
> I'm not sure. I'm thinking of https://github.com/sangria-graphql/sangria/pull/966/files were some validation is done on the AST directly. Does this help?

Yep, I think so.

Also, thank you to have allowed the workflow to run, I need to fix a few things :)

yanns commented:
```
[error] (sangria-core / Compile / scalafmtCheck) scalafmt: 1 files must be formatted (/home/runner/work/sangria/sangria/modules/core)
```

filosganga commented:
@yanns This is ready to review. However, it is impossible to retain binary compatibility because Sangria uses case classes with default values. I have added the mima filters but I think a major version bump may be needed

> > All the implementing fields must be declare When a type or an interface implement another interface, all the implementing fields must be declared. This is not currently validated even without interfaces implementing interfaces, so can it be addressed in a follow-up PR?
> 
> Yes
Actually, I saw that in (https://github.com/sangria-graphql/sangria/blob/7970fa7be466d7bbd5ac07e699e68d769144bc17/modules/core/src/main/scala/sangria/schema/SchemaValidationRule.scala#L125-L127) Sangria allows to avoid defining all the inherited fields.

I have implemented cycle detection in the AST before being materialised

yanns commented:
Really cool, I'll have a look at this.
Yes, I think I'll make a major version bump.

yanns commented on modules/core/src/main/scala/sangria/schema/AstSchemaMaterializer.scala:
```diff
@@ -296,12 +301,45 @@ class AstSchemaMaterializer[Ctx] private (
+
+      val allInterfaceDefs: Vector[ast.InterfaceTypeDefinition] = (interfaceTypeExtensionDefs
+        .find(ie => ie.name == validating.name)
+        .map(ie => ie.interfaces)
+        .getOrElse(Vector.empty) ++ validating.interfaces).flatMap { i =>
```
nitpick: using `.fold(Vector.empty[NamedType])(ie => ie.interfaces) ` instead of map and getOrElse

filosganga replied:
Fixed

yanns reviewed (approved):
Beautiful PR, with lot of tests! 💯

yanns commented:
> If you do a major version bump, I can remove the apply methods that meant to keep source compatibility or limit the changes.

Good point.
We should still try not to break user code if possible. It's sometimes feasible with default values. But sometimes we need those new apply methods.

yanns commented:
For reference, this is in the spec since https://github.com/graphql/graphql-spec/blob/main/changelogs/October2021.md

filosganga commented:
> Good point. We should still try not to break user code if possible. It's sometimes feasible with default values. But sometimes we need those new apply methods.

Fair point

yanns commented:
released in https://github.com/sangria-graphql/sangria/releases/tag/v4.0.0-RC1
